### ISSUE TITLE
[WinBar] remove winbar when the result of an expression is null

### ISSUE THREAD

**Issue by rebelot** [neovim/neovim#18660]
### Feature already in Vim?

no

### Feature description

when evaluating an expression with `%{%...%}` winbar is not removed if the final value is null. For example

```vim
function Empty()
  return ""
endfunction

set winbar=%{%Empty()%}
```
is different than

```vim
set winbar=
```
for example, a function could check the current filetype/buftype and evaluate an appropriate winbar, if the buffer is a prompt, or terminal, or whatever, the function might return `""` to indicate that the winbar should not be drawn.

**Comment by zeertzjq** [comment]
I don't think it is possible to satisfy everyone here. Some may want winbar to still be shown even if it evaluates to empty, while others may want it to be removed in that case.

**Comment by rebelot** [comment]
an empty winbar could be shown by returning `" "` instead of `""` ?

**Comment by zeertzjq** [comment]
A winbar can contain multiple expression to be evaluated, and the resulting text has to be a string, so `v:false` or `v:null` sound quite strange here.

**Comment by rebelot** [comment]
that's absolutely right, I did not think it well

**Comment by kjssad** [comment]
How about an option akin to `laststatus` or `showtabline` where we can have a local value? For now, I'm using using autocommands to filter out filetypes and setting local values for `winbar` there.

**Comment by rebelot** [comment]
Setting local winbar for every buffer except a few ones seems sub optimal.

The reason for this is that if local winbar is unset, it will fall back to global value, and this would require autocmd hacks of sort for anyone wishing to do something sophisticated (statusline plugins?) and I can already see Reddit posts along the lines of "how do I disable winbar in telescope/nvimteee" (not that I like the plugin centric way vim has taken but it's a fact)

It makes total sense to have designed winbar to be compliant with other global/local settings. However a local `nowinbar` option would be at the very least useful to avoid it falling back to the global value for the few buffers one does not wish to have it enabled

**Comment by bfredl** [comment]
Mot sure we want window layout to change on expression evaluation, but we could provide a **static** value to disable winbar locally like `%{%}` or whatever would make sense with stl syntax.

**Comment by famiu** [comment]
> Mot sure we want window layout to change on expression evaluation, but we could provide a **static** value to disable winbar locally like `%{%}` or whatever would make sense with stl syntax.

Alternatively we could add an option that disables the winbar for the current window

**Comment by bfredl** [comment]
Option to disable an option sounds like an anti-pattern. Id rather just identify a good sentinel value.

**Comment by famiu** [comment]
> Option to disable an option sounds like an anti-pattern. Id rather just identify a good sentinel value.

Fair criticism. But not sure what would be a good sentinel value, and how would said sentinel value affect statusline and tabline?

Plus, we already have `laststatus` and `showtabline` which control when statusline and tabline are shown, so why not one for winbar?

**Comment by famiu** [comment]
> Except that 'winbar' already controls whether it is displayed or not, so i think we should rather make that control complete rather than adding a new option only for one specific corner case. Tab and status line does not have this behavior to start with, so these are not equivalent.

That's fair I suppose. Though I can't think of a good value to disable winbar that doesn't seem too arbitrary.

**Comment by mehalter** [comment]
@famiu I think having the ability to disable the winbar makes sense if you are using sidebars like a symbol viewer or file tree sidebar. I also think having a separate option like laststatus/showtabline provides a nice continuity, but maybe it is overkill if you can just set it to a null value

**Comment by famiu** [comment]
> @famiu I think having the ability to disable the winbar per windows makes sense if you are using sidebars like a symbol viewer or file tree sidebar. I also think having a separate option like laststatus/showtabline provides a nice continuity, but maybe it is overkill if you can just set it to a null value unless that wouldn't work on a per-window basis

can't set it to a null value, winbar is a string option and it can only be set to a string

**Comment by mehalter** [comment]
@famiu ah I misunderstood what was meant before when it was mentioned doing `:set winbar=` vs `:set winmar=""`. I guess what I mean to contribute to this conversation is that I think disabling winbar on a per-window basis is very valuable and in the case of users using sidebars such as file trees/symbol outlines/etc. it seems like a relatively necessary feature to keep things looking aesthetic. Sorry for the noise!

**Comment by MunifTanjim** [comment]
Not sure if it's been considered before... I wish there was a generic way to prevent `global-local` options from falling back to global value for a specific window.
